kuka_rsi_hw_interface/src/kuka_hardware_interface.cpp

```diff
@@ -47,95 +47,186 @@
+  std::vector<hardware_interface::StateInterface> state_interfaces;
+  for (size_t i = 0; i < info_.joints.size(); i++) {
+    state_interfaces.emplace_back(
+      info_.joints[i].name,
+      hardware_interface::HW_IF_POSITION,
+      &hw_states_[i]);
+  }
+  return state_interfaces;
+}
+
+std::vector<hardware_interface::CommandInterface> KukaRSIHardwareInterface::
+export_command_interfaces()
+{
+  std::vector<hardware_interface::CommandInterface> command_interfaces;
+  for (size_t i = 0; i < info_.joints.size(); i++) {
+    command_interfaces.emplace_back(
+      info_.joints[i].name,
+      hardware_interface::HW_IF_POSITION,
+      &hw_commands_[i]);
+  }
+  return command_interfaces;
+}
+
+CallbackReturn KukaRSIHardwareInterface::on_activate(const rclcpp_lifecycle::State &)
+{
+  stop_flag_ = false;
   // Wait for connection from robot
   server_.reset(new UDPServer(rsi_ip_address_, rsi_port_));
-  // TODO(Marton): use any logger
-  std::cout << "Waiting for robot connection\n";
+
+  RCLCPP_INFO(rclcpp::get_logger("KukaRSIHardwareInterface"), "Connecting to robot . . .");
+
   int bytes = server_->recv(in_buffer_);
 
+  RCLCPP_INFO(rclcpp::get_logger("KukaRSIHardwareInterface"), "Got data from robot");
+
   // Drop empty <rob> frame with RSI <= 2.3
   if (bytes < 100) {
     bytes = server_->recv(in_buffer_);
   }
 
   rsi_state_ = RSIState(in_buffer_);
-  for (std::size_t i = 0; i < n_dof_; ++i) {
-    joint_state_msg_position[i] = rsi_state_.positions[i] * KukaHardwareInterface::D2R;
-    initial_joint_pos_[i] = rsi_state_.initial_positions[i] * KukaHardwareInterface::D2R;
+
+  for (size_t i = 0; i < info_.joints.size(); ++i) {
+    hw_states_[i] = rsi_state_.positions[i] * KukaRSIHardwareInterface::D2R;
+    hw_commands_[i] = hw_states_[i];
+    initial_joint_pos_[i] = rsi_state_.initial_positions[i] * KukaRSIHardwareInterface::D2R;
   }
   ipoc_ = rsi_state_.ipoc;
-  out_buffer_ = RSICommand(std::vector<double>(n_dof_, 0), ipoc_).xml_doc;
+
+  out_buffer_ = RSICommand(joint_pos_correction_deg_, ipoc_, stop_flag_).xml_doc;
   server_->send(out_buffer_);
-  // Set receive timeout to 1 second
-  server_->set_timeout(1000);
-  // TODO(Marton): use any logger
-  std::cout << "Got connection from robot\n";
+  server_->set_timeout(1000);  // Set receive timeout to 1 second
+
+  RCLCPP_INFO(rclcpp::get_logger("KukaRSIHardwareInterface"), "System Successfully started!");
   is_active_ = true;
+
+  return CallbackReturn::SUCCESS;
+}
+
+CallbackReturn KukaRSIHardwareInterface::on_deactivate(
+  const rclcpp_lifecycle::State &)
+{
+  stop_flag_ = true;
+  RCLCPP_INFO(rclcpp::get_logger("KukaRSIHardwareInterface"), "Stop flag was set!");
+  return CallbackReturn::SUCCESS;
 }
 
-bool KukaHardwareInterface::read(std::vector<double> & joint_state_msg_position)
+return_type KukaRSIHardwareInterface::read(
+  const rclcpp::Time &,
+  const rclcpp::Duration &)
 {
-  std::lock_guard<std::mutex> lock(m_);
   if (!is_active_) {
```

Comment:
if the driver is not active, a few ms wait should be inserted so that the control loop does not block everything
But as I see this sleep is missing 😆 
I will remove isActive() but keep the flag and insert the necessary sleep in inactive case

Reply:
Can this be solved with a timeouting receive? We might miss the first cycle, but at least the code won't have random sleeps

Reply:
you mean do the same recv in inactive state with a big timeout?
if we also have a recv in on_activate() (which we should IMO, see my comment below), than this would lead to trying to receive from two different functions for the same message which might cause undefined behaviour